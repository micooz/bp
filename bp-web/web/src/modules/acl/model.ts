import { ControllerBaseProxy } from 'bizify';
import { configService } from '../../services/config.service';

type Data = {
  file_path: string;
  content: string;
};

export class AclCtrl extends ControllerBaseProxy<Data> {
  $data(): Data {
    return {
      file_path: '',
      content: '',
    };
  }

  services = {
    query: this.$buildService<{ file_path: string; content: string }>(configService.query_acl.bind(configService)),
    modify: this.$buildService<void>(configService.modify.bind(configService)),
  };

  init = async () => {
    const { file_path, content } = await this.services.query.execute();
    this.data.file_path = file_path;
    this.data.content = content;
  };

  handleInputChange = (value: string) => {
    this.data.content = value;
  };

  handleSave = async () => {
    await this.services.modify.execute({
      modify_type: 'acl',
      content: this.data.content,
    });
  };
}
