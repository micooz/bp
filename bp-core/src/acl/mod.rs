mod acl;
mod rule;

pub use acl::*;
pub use rule::*;
