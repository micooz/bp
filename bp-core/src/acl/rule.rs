use std::fmt::{Display, Formatter};

#[derive(Debug, Clone)]
pub struct Rule {
    pub raw: String,
    pub group: RuleGroup,
    pub prefix: RulePrefix,
    pub value: RuleValue,
}

impl Rule {
    pub fn is_allow(&self) -> bool {
        self.group == RuleGroup::Allow
    }
    pub fn is_deny(&self) -> bool {
        self.group == RuleGroup::Deny
    }
}

#[derive(Debug, Clone)]
pub struct RuleValue {
    pub host: String,
    pub port: String,
}

impl RuleValue {
    pub fn is_match(&self, host: &str, port: Option<u16>) -> bool {
        if self.host != "*" && self.host != host {
            return false;
        }
        self.match_port(port)
    }

    pub fn is_fuzzy_match(&self, host: &str, port: Option<u16>) -> bool {
        if self.host != "*" && !host.contains(&self.host) {
            return false;
        }
        self.match_port(port)
    }

    fn match_port(&self, port: Option<u16>) -> bool {
        if self.port == "*" {
            return true;
        }
        if port.is_some() && self.port == port.unwrap().to_string() {
            return true;
        }
        false
    }
}

impl Display for RuleValue {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}:{}",
            if self.host.is_empty() { "*" } else { &self.host },
            if self.port.is_empty() { "*" } else { &self.port },
        )
    }
}

impl From<&str> for RuleValue {
    fn from(s: &str) -> Self {
        let mut split = s.split(':');
        let host = split.next().unwrap_or("*");
        let port = split.next().unwrap_or("*");

        RuleValue {
            host: host.to_string(),
            port: port.to_string(),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum RulePrefix {
    Exact,
    Fuzzy,
    Ignore,
}

impl Display for RulePrefix {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let v = match self {
            Self::Exact => "",
            Self::Fuzzy => "~",
            Self::Ignore => "#",
        };
        write!(f, "{}", v)
    }
}

impl Default for RulePrefix {
    fn default() -> Self {
        Self::Ignore
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum RuleGroup {
    Allow,
    Deny,
}

impl Display for RuleGroup {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let v = match self {
            Self::Allow => "[Allow]",
            Self::Deny => "[Deny]",
        };
        write!(f, "{}", v)
    }
}
