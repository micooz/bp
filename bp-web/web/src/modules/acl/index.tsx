import { useController } from 'bizify';
import { Caption, Button, ErrorBlock, TextArea } from '../../components';
import { useMount } from '../../hooks';
import { AclCtrl } from './model';

export const Acl: React.FC<{}> = () => {
  const vm = useController<AclCtrl>(AclCtrl);
  const { data: vmData, services } = vm;

  useMount(vm.init);

  return (
    <div className="acl">
      <ErrorBlock className="mb-3">{services.query.error?.message}</ErrorBlock>
      <Caption
        extra={<Button size="small" loading={services.modify.loading} onClick={vm.handleSave}>Save</Button>}
        description={<ErrorBlock>{services.modify.error?.message}</ErrorBlock>}
      >
        {vmData.file_path || 'ACL'}
      </Caption>
      <TextArea
        className="logs-textarea"
        style={{ height: '65vh' }}
        value={vmData.content}
        onChange={vm.handleInputChange}
      />
    </div>
  );
};
