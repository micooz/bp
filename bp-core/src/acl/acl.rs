use std::{
    fmt::Display,
    fs,
    io::Write,
    path::PathBuf,
    sync::{mpsc, Arc},
    time,
};

use anyhow::Result;
use notify::{RecommendedWatcher, RecursiveMode, Watcher};
use parking_lot::Mutex;
use serde::Serialize;
use tinytemplate::{format_unescaped, TinyTemplate};

use super::rule::{Rule, RuleGroup, RulePrefix, RuleValue};
use crate::{global, Shutdown};

pub fn get_acl() -> Arc<AccessControlList> {
    global::get_acl()
}

#[derive(Default, Debug)]
pub struct AccessControlList {
    rules: Mutex<Vec<Rule>>,
}

impl AccessControlList {
    pub fn load_from_file(&self, path: &str) -> Result<()> {
        log::info!("loading acl from {}", path);

        let content = fs::read_to_string(path)?;
        self.clear();
        self.deserialize(&content);

        log::info!("loaded {} valid rules", self.count());

        Ok(())
    }

    pub fn save_to_file(&self, path: PathBuf) -> Result<()> {
        let mut file = fs::OpenOptions::new().write(true).create(true).open(path)?;
        let content = self.serialize();

        file.write_all(content.as_bytes())?;
        file.flush()?;

        Ok(())
    }

    pub fn push<V>(&self, group: RuleGroup, prefix: RulePrefix, value: V)
    where
        V: Into<RuleValue> + Display,
    {
        self.rules.lock().insert(
            0,
            Rule {
                raw: format!("{}{}", prefix, value),
                group,
                prefix,
                value: value.into(),
            },
        );
    }

    pub fn try_match(&self, host: &str, port: Option<u16>) -> Option<Rule> {
        let rules = self.rules.lock();

        for rule in rules.iter().rev() {
            match rule.prefix {
                RulePrefix::Exact => {
                    if rule.value.is_match(host, port) {
                        return Some(rule.clone());
                    }
                }
                RulePrefix::Fuzzy => {
                    if rule.value.is_fuzzy_match(host, port) {
                        return Some(rule.clone());
                    }
                }
                RulePrefix::Ignore => {}
            }
        }

        None
    }

    pub fn watch(&self, path: &str, shutdown: Shutdown) -> notify::Result<()> {
        // Create a channel to receive the events.
        let (tx, rx) = mpsc::channel();

        // Automatically select the best implementation for your platform.
        // You can also access each implementation directly e.g. INotifyWatcher.
        let mut watcher: RecommendedWatcher = Watcher::new(tx, time::Duration::from_secs(2))?;

        // Add a path to be watched. All files and directories at that path and
        // below will be monitored for changes.
        watcher.watch(path, RecursiveMode::NonRecursive)?;

        // unwatch when shutdown
        tokio::spawn(async move {
            shutdown.recv().await;
            drop(watcher);
        });

        // This is a simple loop, but you may want to use more complex logic here,
        // for example to handle I/O.
        loop {
            match rx.recv() {
                Ok(notify::DebouncedEvent::Write(_)) => {
                    if let Err(res) = self.load_from_file(path) {
                        log::warn!("reload failed due to: {}", res.to_string())
                    }
                }
                Err(_) => {
                    break;
                }
                _ => continue,
            }
        }

        Ok(())
    }

    pub fn count(&self) -> usize {
        self.rules
            .lock()
            .iter()
            .filter(|&x| x.prefix != RulePrefix::Ignore)
            .count()
    }

    pub fn to_pac(&self, proxy_addr: &str) -> Result<String> {
        let mut if_statements: Vec<String> = vec![];

        for rule in self.rules.lock().iter().rev() {
            let mut if_conditions: Vec<String> = vec![];

            let host_condition = match rule.value.host.as_str() {
                "*" => "".to_string(),
                v => match rule.prefix {
                    RulePrefix::Exact => format!(r#"host === "{}""#, v),
                    RulePrefix::Fuzzy => format!(r#"shExpMatch(host, "*{}*")"#, v),
                    RulePrefix::Ignore => "".to_string(),
                },
            };

            let port_condition = match rule.value.port.as_str() {
                "*" => "".to_string(),
                // port is a js variable defined in assets/pac.tpl
                v => format!("port === {}", v),
            };

            if !host_condition.is_empty() {
                if_conditions.push(host_condition);
            }
            if !port_condition.is_empty() {
                if_conditions.push(port_condition);
            }

            let if_return = match rule.group {
                RuleGroup::Allow => format!("PROXY {}; DIRECT", proxy_addr),
                RuleGroup::Deny => "DIRECT".to_string(),
            };

            let statement = match rule.prefix {
                RulePrefix::Exact | RulePrefix::Fuzzy => {
                    let if_condition = if if_conditions.is_empty() {
                        "true".to_string()
                    } else {
                        if_conditions.join(" && ")
                    };
                    format!(r#"if ({}) return "{}";"#, if_condition, if_return)
                }
                RulePrefix::Ignore => {
                    format!("// {}", rule.raw)
                }
            };

            // add two leading spaces for better readability
            if_statements.push(format!("  {}", statement));
        }

        let mut tt = TinyTemplate::new();
        tt.add_template("pac", include_str!("assets/pac.tpl"))?;
        tt.set_default_formatter(&format_unescaped);

        #[derive(Serialize)]
        struct Context {
            if_statements: String,
        }

        let rendered = tt.render(
            "pac",
            &Context {
                if_statements: if_statements.join("\n"),
            },
        )?;

        Ok(rendered)
    }

    fn clear(&self) {
        self.rules.lock().clear();
    }

    fn deserialize(&self, content: &str) {
        let mut rules = self.rules.lock();
        let mut group = RuleGroup::Deny;

        for line in content.lines() {
            let line = line.trim();

            // ignore empty line
            if line.is_empty() {
                continue;
            }

            // only keep the first part, for example: "keep_me ignore_me also_ignore" -> "keep_me"
            let mut split = line.split(' ');
            let item = split.next().unwrap();

            // determine to use which group
            match item.to_uppercase().as_str() {
                "[ALLOW]" => {
                    group = RuleGroup::Allow;
                    continue;
                }
                "[DENY]" => {
                    group = RuleGroup::Deny;
                    continue;
                }
                _ => {}
            }

            // obtain prefix
            let prefix = match item.chars().next().unwrap() {
                '~' => RulePrefix::Fuzzy,
                '#' => RulePrefix::Ignore,
                _ => RulePrefix::Exact,
            };

            // skip prefix
            let skip_n = usize::from(prefix != RulePrefix::Exact);
            let chars = item.chars().skip(skip_n);

            // to string
            let value = chars.collect::<String>();

            rules.push(Rule {
                raw: line.to_string(),
                group: group.clone(),
                prefix,
                value: value.as_str().into(),
            });
        }
    }

    fn serialize(&self) -> String {
        let rules = self.rules.lock();
        let groups = rules.group_by(|a, b| a.group == b.group);
        let mut lines: Vec<String> = vec![];

        for group in groups.rev() {
            let name = group[0].group.to_string();
            lines.push(name);
            lines.push("\n".to_string());

            for rule in group.iter().rev() {
                lines.push(format!("{}{}\n", rule.prefix, rule.value));
            }
            lines.push("\n".to_string());
        }

        lines.join("")
    }
}
