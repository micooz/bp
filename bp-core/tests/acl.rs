use std::fs::read_to_string;

use bp_core::acl::{AccessControlList, RuleGroup, RulePrefix};

#[test]
fn test_load_from_file() {
    let acl = AccessControlList::default();
    assert!(acl.load_from_file("tests/fixtures/acl_simple.txt").is_ok());
    assert_eq!(acl.count(), 7);
}

#[test]
fn test_to_pac() {
    let acl = AccessControlList::default();
    assert!(acl.load_from_file("tests/fixtures/acl_simple.txt").is_ok());
    insta::assert_snapshot!(acl.to_pac("127.0.0.1:1080").unwrap());
}

#[test]
fn test_save_to_file() {
    let acl = AccessControlList::default();
    acl.push(RuleGroup::Allow, RulePrefix::Exact, "");
    acl.push(RuleGroup::Allow, RulePrefix::Exact, "*");
    acl.push(RuleGroup::Allow, RulePrefix::Exact, "*:");
    acl.push(RuleGroup::Allow, RulePrefix::Exact, "*:*");
    acl.push(RuleGroup::Allow, RulePrefix::Fuzzy, "");
    acl.push(RuleGroup::Allow, RulePrefix::Fuzzy, "*");

    acl.push(RuleGroup::Deny, RulePrefix::Exact, "bar1.com");
    acl.push(RuleGroup::Deny, RulePrefix::Fuzzy, "bar2.com:80");
    acl.push(RuleGroup::Deny, RulePrefix::Ignore, "bar3.com:443");

    acl.push(RuleGroup::Allow, RulePrefix::Exact, "foo1.com");
    acl.push(RuleGroup::Allow, RulePrefix::Fuzzy, "foo2.com:80");
    acl.push(RuleGroup::Allow, RulePrefix::Ignore, "foo3.com:443");

    let tmp_path = "tests/tmp/acl.txt";

    assert!(acl.save_to_file(tmp_path.into()).is_ok());
    insta::assert_snapshot!(read_to_string(tmp_path).unwrap());
}

#[test]
fn test_try_match() {
    let acl = AccessControlList::default();

    acl.load_from_file("tests/fixtures/acl_simple.txt").unwrap();
    assert!(acl.try_match("xxx.com", None).is_none());
    assert!(acl.try_match("example.net", None).unwrap().is_deny());
    assert!(acl.try_match("example.net", Some(80)).unwrap().is_deny());
    assert!(acl.try_match("foo_baz.com", None).unwrap().is_deny());

    acl.load_from_file("tests/fixtures/acl_whitelist.txt").unwrap();
    assert!(acl.try_match("example.com", None).unwrap().is_allow());
    assert!(acl.try_match("example.com", Some(53)).unwrap().is_allow());
    assert!(acl.try_match("_example.com", None).unwrap().is_deny());
    assert!(acl.try_match("example.net", None).unwrap().is_deny());

    acl.load_from_file("tests/fixtures/acl_blacklist.txt").unwrap();
    assert!(acl.try_match("example.com", None).unwrap().is_deny());
    assert!(acl.try_match("example.com", Some(53)).unwrap().is_deny());
    assert!(acl.try_match("_example.com", None).unwrap().is_allow());
    assert!(acl.try_match("example.net", None).unwrap().is_allow());

    acl.load_from_file("tests/fixtures/acl_mixed.txt").unwrap();
    assert!(acl.try_match("baz.com", Some(443)).unwrap().is_deny());
    assert!(acl.try_match("example.com", Some(443)).unwrap().is_allow());
    assert!(acl.try_match("bar.com", None).unwrap().is_allow());
    assert!(acl.try_match("deny.com", None).unwrap().is_deny());
}
