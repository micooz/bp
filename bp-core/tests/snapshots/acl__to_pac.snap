---
source: bp-core/tests/acl.rs
assertion_line: 16
expression: "acl.to_pac(\"127.0.0.1:1080\").unwrap()"

---
// This file is automatically generated.

function FindProxyForURL(url, host) {
  // extract port number from url
  var port_match = url.match(/:(\d{0,5})/);
  var port = port_match[1] ? Number(port_match[1]) : null;

  if (host === "allow.com" && port === 8080) return "PROXY 127.0.0.1:1080";
  if (host === "example.net") return "DIRECT";
  if (port === 443) return "DIRECT";
  if (host === "apple.com" && port === 80) return "DIRECT";
  if (shExpMatch(host, "*baz.com*")) return "DIRECT";
  // #empty rule
  // #full_line comment
  if (host === "bar.com") return "DIRECT";
  if (host === "foo.com") return "DIRECT";

  return "DIRECT";
}

